File: Web/autoCheckMobileReact/src/shared/lib/appLogger.ts
type LogLevel = 'DEBUG' | 'INFO' | 'ERROR'

function normalizeDetails(details?: unknown) {
  if (details == null) {
    return 'Без дополнительных данных'
  }
  if (typeof details === 'string') {
    return details
  }
  if (details instanceof Error) {
    return details.message
  }
  try {
    return JSON.stringify(details)
  } catch {
    return String(details)
  }
}

function write(level: LogLevel, component: string, event: string, details?: unknown) {
  const message = `[${component}]: ${level} ${event} — ${normalizeDetails(details)}`
  if (level === 'ERROR') {
    console.error(message)
    return
  }
  if (level === 'DEBUG') {
    console.debug(message)
    return
  }
  console.info(message)
}

/**
 * Назначение: единый frontend logger в формате, заданном конкурсным заданием.
 * Дата создания: 31-05-2026.
 * Автор: Команда.
 */
export const appLogger = {
  debug: (component: string, event: string, details?: unknown) => write('DEBUG', component, event, details),
  info: (component: string, event: string, details?: unknown) => write('INFO', component, event, details),
  error: (component: string, event: string, details?: unknown) => write('ERROR', component, event, details),
}

File: Mobile/autoCheckMobileFlutter/lib/src/services/app_logger.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';

/// Назначение: единая точка клиентского логирования Flutter-приложения.
///
/// Ответственность/SRP: форматирует события экранов, репозитория и пользовательских
/// действий в виде `[Component]: LEVEL Event - Details`, чтобы логи совпадали с
/// конкурсным требованием DEBUG/INFO/ERROR.
///
/// Дата создания: 31-05-2026.
/// Автор: Команда.
///
/// Публичные методы:
/// - [debug] — пишет диагностические события и успешные API-ответы.
/// - [info] — пишет жизненный цикл экранов и критичные действия пользователя.
/// - [error] — пишет ошибки API, валидации и выполнения сценариев.
class AppLogger {
  const AppLogger._();

  static void debug(String component, String event, [Object? details]) {
    _write('DEBUG', component, event, details);
  }

  static void info(String component, String event, [Object? details]) {
    _write('INFO', component, event, details);
  }

  static void error(String component, String event, [Object? details]) {
    _write('ERROR', component, event, details);
  }

  static void _write(
    String level,
    String component,
    String event,
    Object? details,
  ) {
    final suffix = _normalize(details);
    debugPrint('[$component]: $level $event - $suffix');
  }

  static String _normalize(Object? details) {
    if (details == null) return 'No details';
    if (details is Map || details is List) {
      try {
        return jsonEncode(details);
      } catch (_) {
        return details.toString();
      }
    }
    return details.toString();
  }
}
